framework/smoke_test_framework.cpp

```diff
@@ -0,0 +1,592 @@
+    }
+
+    Result<JSON::String> _revocableTaskResources =
+        json.get().find<JSON::String>(
+            "tasks[" + stringify(i) +"].revocableResources");
+
+    if (_revocableTaskResources.isSome()) {
+      revocable = true;
+      Try<Resources> _revocableResources =
+          Resources::parse(_revocableTaskResources.get().value);
+
+      if (_revocableResources.isError()) {
+        EXIT(EXIT_FAILURE)
+        << flags.usage("Invalid revocableResources in JSON: " +
+                       _revocableResources.error());
+      }
+
+          foreach (Resource _revocable, _revocableResources.get()) {
+              _revocable.mutable_revocable();
+              _resources.get() += _revocable;
+            }
+    }
+
+    Option<string> _targetHostname = None();
+
+    Result<JSON::String> _target =
+        json.get().find<JSON::String>(
+            "tasks[" + stringify(i) +"].targetHostname");
+
+    if (_target.isSome()) {
+      _targetHostname = _target.get().value;
+    }
+
+    Option<size_t> _taskNum = None();
+
+    Result<JSON::Number> _totalTasks =
+        json.get().find<JSON::Number>(
+            "tasks[" + stringify(i) +"].totalTasks");
+
+    if (_totalTasks.isSome()) {
+      _taskNum = _totalTasks.get().value;
+    }
+
+    jobs.push_back(JobSpec(_command.get().value,
+                           _resources.get(),
+                           _taskNum,
+                           _targetHostname));
+    jobs.back().print();
+  }
+}
+
+int main(int argc, char** argv)
+{
+  bool enableRevocable = false;
+  Flags flags;
+
+  Try<Nothing> load = flags.load("MESOS_", argc, argv);
+
+  if (load.isError()) {
+    EXIT(EXIT_FAILURE)
+      << flags.usage(load.error());
+  }
+
+  if (flags.help) {
+    EXIT(EXIT_SUCCESS)
+      << flags.usage();
+  }
+
+  if (flags.master.isNone()) {
+    EXIT(EXIT_FAILURE)
+      << flags.usage("Missing required option --master");
+  }
+
+  if (flags.principal.isSome() != flags.secret.isSome()) {
+    EXIT(EXIT_FAILURE)
+      << flags.usage("Both --principal and --secret are required"
+                     " to enable authentication");
+  }
+
+  logging::initialize(argv[0], flags, true); // Catch signals.
+
+  FrameworkInfo framework;
+  framework.set_user(""); // Have Mesos fill in the current user.
+  framework.set_name("Serenity Smoke Test Framework");
+  framework.set_checkpoint(flags.checkpoint);
+  framework.set_role(flags.role);
+
+  list<JobSpec> jobs;
+  if (flags.tasks_json_path.isSome())
+  {
+    parseTaskJson(jobs, flags, enableRevocable);
+
+  } else {
+    // Task specification.
+    Try<Resources> resources =
+        Resources::parse(flags.task_resources);
+
+    if (resources.isError()) {
+      EXIT(EXIT_FAILURE)
+      << flags.usage("Invalid --task_resources: " +
+                     resources.error());
+    }
+
+    Resources taskResources = resources.get();
+
+    if (flags.task_revocable_resources.isSome()) {
+      enableRevocable = true;
```

Comment:
Does it matter? In current implementation we prefer task json specification (it has higher priority).